storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+	if err != nil {
+		logrus.Errorf("error reading image %q: %v", imageRef.ID(), err)
+		return nil, err
+	}
+	tag = imageRef.StringWithinTransport()
+	image := storageImage{
+		store:    imageRef.store,
+		imageRef: imageRef,
+		Tag:      tag,
+		Created:  time.Now(),
+		ID:       img.ID,
+		BlobList: []string{},
+		Layers:   make(map[string]string),
+		BlobData: make(map[string][]byte),
+	}
+	if err := image.loadMetadata(); err != nil {
+		logrus.Errorf("error decoding metadata for source image: %v", err)
+		return nil, err
+	}
+	return &image, nil
+}
+
+// newImageDestination sets us up to write a new image.
+func newImageDestination(imageRef *storageReference) (*storageImage, error) {
+	// We set the image's ID if the reference we got looked like one, since
+	// we take that as an indication that it's going to end up with the
+	// same contents as an image we already have.  If the reference looks
+	// more like a name, we don't know yet if it'll be exactly the same as,
+	// or be an updated version of, an image we might already have, so we
+	// have to err on the side of caution and create a new image which will
+	// be assigned the name as its tag.
+	image := storageImage{
+		store:    imageRef.store,
+		imageRef: imageRef,
+		Tag:      imageRef.StringWithinTransport(),
+		Created:  time.Now(),
+		ID:       imageRef.ID(),
+		BlobList: []string{},
+		Layers:   make(map[string]string),
+		BlobData: make(map[string][]byte),
+	}
+	return &image, nil
+}
+
+func newImage(imageRef *storageReference) (*storageImage, error) {
+	return newImageSource(imageRef)
+}
+
+func (s *storageImage) loadMetadata() error {
+	if s.ID != "" {
+		image, err := s.store.GetImage(s.ID)
+		if image != nil && err == nil {
+			if err := json.Unmarshal([]byte(image.Metadata), s); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *storageImage) saveMetadata() error {
+	if s.ID != "" {
+		metadata, err := json.Marshal(s)
+		if len(metadata) != 0 && err == nil {
+			istore, err := s.store.GetImageStore()
+			if istore != nil && err == nil {
+				if err = istore.SetMetadata(s.ID, string(metadata)); err != nil {
+					logrus.Errorf("error setting metadata for image %q: %v", s.ID, err)
+				}
+			} else {
+				logrus.Errorf("error locating image store: %v", err)
+			}
+			return err
+		}
+		logrus.Errorf("error encoding metadata for image %q: %v", s.ID, err)
+		return err
+	}
+	return nil
+}
+
+func (s *storageImage) Reference() types.ImageReference {
+	return s.imageRef
+}
+
+func (s *storageImage) Close() {
+}
+
+// SupportsSignatures returns an error if we can't expect GetSignatures() to
+// return data that was previously supplied to PutSignatures().
+func (s *storageImage) SupportsSignatures() error {
+	return nil
+}
+
+// PutBlob is used to both store filesystem layers and binary data that is part
+// of the image.  Filesystem layers are assumed to be imported in order, as
+// that is required by some of the underlying storage drivers.
```

Comment:
Then it's not providing the storage `ImageDestination` with enough information to correctly set up a filesystem tree with the image's contents, since the blobs are not purely additions, but deltas.

If we replace a duplicate call to `PutBlob` with something like a `PutDuplicatedBlob`, or give `PutBlob` a flag that indicates that its `io.Reader` argument isn't going to be providing data because the image should already have a copy, we can do the right thing.  But some indication is needed.